rfcs/0062-content-addressed-paths.md

```diff
@@ -54,38 +54,71 @@ cutoffs.
 
 [design]: #detailed-design
 
+*In everything that follows, most algorithms and data-structures will be expressed as pseudo-python snippets*
+
 When it comes to computing the output paths of a derivation, the current Nix
 model, known as the “input-addressed” model (also sometimes referred to as the
 “extensional” model) works (roughly) as follows:
 
-- A Derivation is a data-structure that specifies how to build a package.
-  Derivations can refer to other derivations
-- All these derivations have a “hash-modulo” associated to them, which is defined by:
-  - Some derivations known as “fixed-output” have a known result (for example
-    because they fetch a tarball from the internet, and we assume that this
-    tarball will stay immutable).
-    These have their output hash manually defined (and this hash will be
-    checked against the actual hash of their output when they get built)
-  - All the others have a hash that's recursively computed by the following algorithm:
-    - If a derivation doesn't depend on any other derivation, then we just hash its representation,
-    - Otherwise, we substitute each occurence of a dependency by its hash modulo and hash the result.
-- For each output of a derivation, we compute the associated output path by
-  hashing the hash modulo of the derivation and the output name.
-
-This proposal adds a new kind of derivation: “floating content-addressed
-derivations”, which are similar to fixed-output derivations in that they are
-stored in a content-addressed path, but don't have this output hash specified
-ahead of time.
-
-For this to work properly, we need to extend the current build process, as well
-as the caching and remote building systems so that they are able to take into
-account the specificies of these new derivations.
-
-## Nix-build process
-
-For the sake of clarity, we will refer to the current model (where the
-derivations are indexed by their inputs, also sometimes called "extensional") as
-the `input-addressed` model
+1. A Nix language expression gets evaluated to a `derivation`
+2. This `derivation` is a data-structure describing how to build a package. In particular it contains
+  1. A set of derivation outputs which will be used as input for the build
```

Comment:
Yes, there’s a bit of nesting here that makes things a bit tricky to explain. It refers to the `inputDrvs` key (which is a set of derivation associated with a list of output names. Not sure how to word this better, any suggestion is wecome :) ).

Reply:
The problematic part is just that it wasn't clear to me at first glance what's the difference between 1. and 2. (what is the difference between a derivation output vs. store path). I suppose this might also be a bit ambiguous even when taking nixphd as the main reference...

> Figure 2.12 outlines this two-stage build process: Nix expressions are first translated to store derivations that live in the Nix store and that each describe a single build action with all variability removed. These store derivations can then be built, which results in derivation outputs that also live in the Nix store.

(nixphd, p. 39)

> The idea is that pre-built derivation outputs are made available in some central repository accessible to the client machines

(nixphd, p. 45)